core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala

```diff
@@ -65,48 +65,84 @@ case class MetadataSnapshot(partitionStates: mutable.AnyRefMap[String, mutable.L
 }
 
 object ZkMetadataCache {
-  /**
-   * Create topic deletions (leader=-2) for topics that are missing in a FULL UpdateMetadataRequest coming from a
-   * KRaft controller during a ZK migration. This will modify the UpdateMetadataRequest object passed into this method.
-   */
-  def maybeInjectDeletedPartitionsFromFullMetadataRequest(
+  def transformKRaftControllerFullMetadataRequest(
     currentMetadata: MetadataSnapshot,
     requestControllerEpoch: Int,
     requestTopicStates: util.List[UpdateMetadataTopicState],
-  ): Seq[Uuid] = {
-    val prevTopicIds = currentMetadata.topicIds.values.toSet
-    val requestTopics = requestTopicStates.asScala.map { topicState =>
-      topicState.topicName() -> topicState.topicId()
-    }.toMap
-
-    val deleteTopics = prevTopicIds -- requestTopics.values.toSet
-    if (deleteTopics.isEmpty) {
-      return Seq.empty
+  ): (util.List[UpdateMetadataTopicState], util.List[String]) = {
+    val topicIdToNewState = new util.HashMap[Uuid, UpdateMetadataTopicState]()
+    requestTopicStates.forEach(state => topicIdToNewState.put(state.topicId(), state))
+    val logMessages = new util.ArrayList[String]
+    val newRequestTopicStates = new util.ArrayList[UpdateMetadataTopicState]()
+    currentMetadata.topicNames.forKeyValue((id, name) => {
+      Option(topicIdToNewState.get(id)) match {
+        case None =>
+          currentMetadata.partitionStates.get(name) match {
+            case None => logMessages.add(s"Error: topic ${name} appeared in currentMetadata.topicNames, " +
+              "but not in currentMetadata.partitionStates.")
+            case Some(oldPartitionStates) =>
```

Comment:
I feel like "new" is pretty clear, but I agree that we should choose one of "current" / "old" and stick to it. Since the code already uses "current" all over the place I'll go with that.

Reply:
In `ZkMetadataCache.updateMetadata`, I'll also replace "oldData" with "originalRequestData" since that is clearer.